mobilecoind/src/sync.rs

```diff
@@ -345,69 +346,77 @@ fn match_tx_outs_into_utxos(
-
-        let (amount, _blinding) = tx_out
-            .get_masked_amount()
-            .expect("missing masked amount")
-            .get_value(&shared_secret)
-            .expect("Malformed amount"); // TODO
-
-        let onetime_private_key = recover_onetime_private_key(
-            &tx_public_key,
-            account_key.view_private_key(),
-            &account_key.subaddress_spend_private(subaddress_id.index),
-        );
-
-        let key_image = KeyImage::from(&onetime_private_key);
-
-        results.push(UnspentTxOut {
-            tx_out: tx_out.clone(),
-            subaddress_index: subaddress_id.index,
-            key_image,
-            value: amount.value,
-            attempted_spend_height: 0,
-            attempted_spend_tombstone: 0,
-            token_id: *amount.token_id,
-        });
-    }
+    // Iterate over each output and filter the results using a parallel iterator.
+    let results: Result<Vec<UnspentTxOut>, Error> = outputs
+        .into_par_iter()
+        .filter_map(|tx_out| {
+            // Convert target and public keys to RistrettoPublic type.
+            let tx_out_target_key = RistrettoPublic::try_from(&tx_out.target_key).ok()?;
+            let tx_public_key = RistrettoPublic::try_from(&tx_out.public_key).ok()?;
+
+            // Generate subaddress spend public key for tx_out.
+            let subaddress_spk = SubaddressSPKId::from(&recover_public_subaddress_spend_key(
+                account_key.view_private_key(),
+                &tx_out_target_key,
+                &tx_public_key,
+            ));
+
+            // Search the database for the subaddress ID that matches the generated key.
+            let subaddress_id = match mobilecoind_db.get_subaddress_id_by_spk(&subaddress_spk) {
+                Ok(data) => {
+                    // Log the index and monitor ID of the matched subaddress.
+                    log::trace!(
+                        logger,
+                        "matched subaddress index {} for monitor_id {}",
+                        data.index,
+                        data.monitor_id,
+                    );
 
-    Ok(results)
+                    data
+                }
+                Err(Error::SubaddressSPKNotFound) => return None,
+                Err(e) => return Some(Err(e)),
+            };
+
+            // Check that the matched subaddress belongs to the current monitor.
+            assert_eq!(monitor_id, &subaddress_id.monitor_id);
+
+            // Generate the shared secret between the subaddress and output public key.
+            let shared_secret =
+                get_tx_out_shared_secret(account_key.view_private_key(), &tx_public_key);
+
+            // Get the amount and blinding factor for the output.
+            let (amount, _blinding) = tx_out
+                .get_masked_amount()
+                .expect("missing masked amount")
+                .get_value(&shared_secret)
+                .expect("Malformed amount"); // TODO
```

Comment:
What's the TODO here? Can we also put it above the line rather than to the right if it's still needed?